### ISSUE TITLE
Add a RandomBatchSampler ?

### ISSUE THREAD

**Issue by name-yy** [pytorch/pytorch#18317]
## 🚀 Feature
Shuffle samples by batch

## Motivation
I'm working on a variable length sequence classification problem and use collate_fn to padding zeros in each batch. Sorting the samples approximates the length of the samples within a batch to avoid excessive padding. So I would prefer to shuffle only by batch rather than by sample.

## Pitch

Shuffle by batch in each epoch

## Alternatives
I write a RandomBatchSampler that works well.

```
import torch
from torch.utils.data import Sampler


class RandomBatchSampler(Sampler):
    r"""Yield a mini-batch of indices with random batch order

    Arguments:
        data_source (Dataset): dataset to sample from
        batch_size (int): Size of mini-batch
        drop_last (bool): If ``True``, the sampler will drop the last batch if
            its size would be less than ``batch_size``
    """

    def __init__(self, data_source, batch_size, drop_last):

        if not isinstance(batch_size, int) or batch_size <= 0:
            raise ValueError("batch_size should be a positive integeral "
                             "value, but got batch_size={}".format(batch_size))
        if not isinstance(drop_last, bool):
            raise ValueError("drop_last should be a boolean value, but got "
                             "drop_last={}".format(drop_last))

        self.data_source = data_source
        self.batch_size = batch_size
        self.drop_last = drop_last
        self.fragment_size = len(data_source) % batch_size

    def __iter__(self):
        batch_indices = range(0, len(self.data_source) - self.fragment_size, self.batch_size)

        for batch_indices_idx in torch.randperm(len(self.data_source) // self.batch_size):
            yield list(range(batch_indices[batch_indices_idx], batch_indices[batch_indices_idx]+self.batch_size))

        if self.fragment_size > 0 and not self.drop_last:
            yield list(range(len(self.data_source) - self.fragment_size, len(self.data_source)))

    def __len__(self):
        if self.drop_last:
            return len(self.data_source) // self.batch_size
        else:
            return (len(self.data_source) + self.batch_size - 1) // self.batch_size

```

**Comment by fmassa** [comment]
I believe the implementation as is assumes some structure on the dataset which doesn't always hold true.

Having a generic implementation could be useful, but I believe things like this might be better in torchtext.

cc @cpuhrsch

**Comment by name-yy** [comment]
> I believe the implementation as is assumes some structure on the dataset which doesn't always hold true.
> 
> Having a generic implementation could be useful, but I believe things like this might be better in torchtext.
> 
> cc @cpuhrsch

Yes, this implementation assumes the dataset is sorted. Actually, I sort the dataset before the "DataSet" and the "DataLoader".

**Comment by cpuhrsch** [comment]
Adding data processing capabilities in the context of torchtext has not made it onto the mid-term roadmap. Therefore, I don't have much to contribute to this discussion at this point.

**Comment by ejguan** [comment]
torchtext has already implemented a BucketIterator to support this feature. Please take a look at the example datasets in torchtext for reference (https://github.com/pytorch/text/tree/main/torchtext/legacy/datasets)

We are also working on the new DataLoader DataPipe system that would provide more flexible approach to run data loading. And, torchtext and our team are working on examples to do this bucket sampling and data loading using DataPipe. We will provide more detailed example shortly.
